frontend/src/utilities/__tests__/useDraggableTable.spec.ts

```diff
@@ -0,0 +1,123 @@
+    row.id = itemId;
+    tbody.appendChild(row);
+  });
+  document.body.appendChild(tbody);
+  return tbody;
+};
+describe('initializeState', () => {
+  it('should update state and classes on drag actions', () => {
+    const itemOrder = ['item1', 'item2', 'item3'];
+    const bodyRef = generateRowsWithItems(itemOrder);
+    const renderResult = testHook(useDraggableTable)(itemOrder, setItemOrder);
+    renderResult.result.current.tableProps.tbodyProps.ref.current = bodyRef;
+
+    // Initial state assertions
+    expect(renderResult).hookToHaveUpdateCount(1);
+    expect(renderResult.result.current.tableProps.className).toBe(undefined);
+
+    // Simulate drag start event
+    const { onDragStart } = renderResult.result.current.rowProps;
+    act(() => {
+      onDragStart({
+        dataTransfer: { effectAllowed: 'none', setData: jest.fn() },
+        currentTarget: {
+          id: 'item3',
+          classList: { add: jest.fn() },
+          setAttribute: jest.fn(),
+        },
+      } as unknown as React.DragEvent<HTMLTableRowElement>);
+    });
+    expect(renderResult).hookToHaveUpdateCount(2);
+    expect(renderResult.result.current.tableProps.className).toBe('pf-m-drag-over');
+
+    // Simulate drag over
+    const { onDragOver } = renderResult.result.current.tableProps.tbodyProps;
+    act(() => {
+      onDragOver({
+        preventDefault: jest.fn(),
+        target: bodyRef.children[0],
+      } as unknown as React.DragEvent<HTMLTableSectionElement>);
+    });
+
+    expect(renderResult).hookToHaveUpdateCount(3);
+    expect(renderResult.result.current.tableProps.className).toBe('pf-m-drag-over');
+
+    // Simulate drag end
+    const { onDragEnd } = renderResult.result.current.rowProps;
+    act(() => {
+      onDragEnd({
+        target: {
+          id: 'item4',
+          classList: { remove: jest.fn() },
+          setAttribute: jest.fn(),
+        },
+      } as unknown as React.DragEvent<HTMLTableRowElement>);
+    });
+
+    expect(renderResult).hookToHaveUpdateCount(4);
+    expect(renderResult.result.current.tableProps.className).toBe(undefined);
+
+    // Simulate drop event
+    const { onDrop } = renderResult.result.current.rowProps;
+    renderResult.result.current.tableProps.tbodyProps.ref.current = bodyRef;
+    renderResult.result.current.tableProps.tbodyProps.ref.current.getBoundingClientRect = jest
+      .fn()
+      .mockReturnValue({
+        x: 40,
+        y: 40,
+        width: 20,
+        height: 20,
+      } as DOMRect);
+
+    act(() => {
+      onDrop({
+        clientX: 45,
+        clientY: 45,
+      } as unknown as React.DragEvent<HTMLTableRowElement>);
+    });
+    expect(setItemOrder).toHaveBeenCalledWith(['item3', 'item1', 'item2']);
+
+    // Test isValidDrop when drop is not valid
+    renderResult.result.current.tableProps.tbodyProps.ref.current.getBoundingClientRect = jest
+      .fn()
+      .mockReturnValue({} as DOMRect);
+    act(() => {
+      onDrop({
```

Comment:
Why would `onDrop` be called twice in a row? You also perform the exact same assertions, accidental duplicate code?

Reply:
`onDrop` called once with a valid drop event and second time with invalid drop event to execute `onDragCancel`